### PR TITLE
Alter view set authorization

### PR DESCRIPTION
**Pull request by skrzypo987**
Rebased on top of https://github.com/prestosql/presto/pull/5717

### CODE REVIEW

**Comment by findepi** [comment]
Is it supposed to work with Hive views?
do we have a test covering this (regardless of the answer to the above question)?

File: presto-parser/src/main/java/io/prestosql/sql/tree/SetAuthorizationStatement.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public abstract class SetAuthorizationStatement
```
Comment (kokosing) [inline presto-parser/src/main/java/io/prestosql/sql/tree/SetAuthorizationStatement.java:25]: How about extracting abstract Task?

**Review by kokosing — COMMENTED** [review]
So far so good

**Comment by kokosing** [comment]
> Is it supposed to work with Hive views?

Good point. I don't see a reason why not. It is just change of ownership, not the body of the view.

This opens a question about other `ALTER VIEW name RENAME TO new_name` statement and Hive views. I don't recall any tests for that either and also it looks like something that should just work.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 2]
```diff
@@ -831,6 +831,70 @@ public void testTableAuthorizationForRole()
...
+                .build();
+
+        Session alice = testSessionBuilder()
+                .setCatalog(getSession().getCatalog().get())
+                .setIdentity(Identity.forUser("alice").build())
+                .build();
+
+        assertUpdate(admin, "CREATE SCHEMA test_table_authorization");
+        assertUpdate(admin, "CREATE TABLE test_table_authorization.foo (col int)");
+        assertUpdate(admin, "CREATE VIEW test_table_authorization.bar AS SELECT * FROM test_table_authorization.foo");
```
[the remark this answers is not in the document]
Comment (skrzypo987) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: I added tests with INVOKER and DEFINER. current_user function does not work like that thought so there is a standard view referencing a table to test that.

File: presto-parser/src/main/java/io/prestosql/sql/tree/SetAuthorizationStatement.java (hunk 1 above)
Comment (skrzypo987) [reply presto-parser/src/main/java/io/prestosql/sql/tree/SetAuthorizationStatement.java:25]: Those tasks actually vary so I don't think it will result in less code.

File: presto-plugin-toolkit/src/test/java/io/prestosql/plugin/base/security/TestFileBasedAccessControl.java [hunk 3]
```diff
@@ -301,6 +301,13 @@ public void testTableRules()
         accessControl.checkCanSetTableAuthorization(ALICE, aliceTable, new PrestoPrincipal(PrincipalType.USER, "some_user"));
         assertDenied(() -> accessControl.checkCanSetTableAuthorization(ALICE, bobTable, new PrestoPrincipal(PrincipalType.ROLE, "some_role")));
         assertDenied(() -> accessControl.checkCanSetTableAuthorization(ALICE, bobTable, new PrestoPrincipal(PrincipalType.USER, "some_user")));
+
+        accessControl.checkCanSetViewAuthorization(ADMIN, testTable, new PrestoPrincipal(PrincipalType.ROLE, "some_role"));
```
Comment (s2lomon) [inline presto-plugin-toolkit/src/test/java/io/prestosql/plugin/base/security/TestFileBasedAccessControl.java:305]: Don't we want to make separate tests for such cases?

File: presto-plugin-toolkit/src/test/java/io/prestosql/plugin/base/security/TestFileBasedSystemAccessControl.java [hunk 4]
```diff
@@ -590,6 +595,8 @@ public void testCheckCanSetTableAuthorizationForNonOwner()
 
         assertAccessDenied(() -> accessControl.checkCanSetTableAuthorization(ALICE, new CatalogSchemaTableName("some-catalog", "test", "test"), new PrestoPrincipal(PrincipalType.ROLE, "some_role")), AUTH_TABLE_ACCESS_DENIED_MESSAGE);
         assertAccessDenied(() -> accessControl.checkCanSetTableAuthorization(ALICE, new CatalogSchemaTableName("some-catalog", "test", "test"), new PrestoPrincipal(PrincipalType.USER, "some_user")), AUTH_TABLE_ACCESS_DENIED_MESSAGE);
+        assertAccessDenied(() -> accessControl.checkCanSetViewAuthorization(ALICE, new CatalogSchemaTableName("some-catalog", "test", "test"), new PrestoPrincipal(PrincipalType.ROLE, "some_role")), AUTH_VIEW_ACCESS_DENIED_MESSAGE);
```
Comment (s2lomon) [inline presto-plugin-toolkit/src/test/java/io/prestosql/plugin/base/security/TestFileBasedSystemAccessControl.java:598]: I understand that for these scenarios the implementations simply checks for access for table, so putting these assertions here works, but It's something that requires the knowledge of the actual implementation to understand this test and therefore I would see it rather as a separate test. Not crucial, but think about it.

**Review by s2lomon — APPROVED** [review]
lgtm

**Comment by skrzypo987** [comment]
@s2lomon I split some tests in one of the classes you mentioned and left as it is in the other.
@kokosing Made changes to tests to check the DEFINER and INVOKER types of views.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 5]
```diff
@@ -853,6 +853,126 @@ public void testTableAuthorizationForRole()
...
+
+        Session alice = testSessionBuilder()
+                .setCatalog(getSession().getCatalog().get())
+                .setIdentity(Identity.forUser("alice").build())
+                .build();
+
+        assertUpdate(admin, "CREATE SCHEMA test_view_authorization");
+        assertUpdate(admin, "CREATE TABLE test_view_authorization.foo (col int)");
+        assertUpdate(admin, "INSERT INTO test_view_authorization.foo VALUES (1)", 1);
+        assertUpdate(admin, "CREATE VIEW test_view_authorization.bar SECURITY DEFINER AS SELECT * from test_view_authorization.foo");
```
Comment (kokosing) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:899]: this is really unreadable to cope with `bar` and `foo`. I need to check couple times what is what when trying to understand error messages below.

Comment (kokosing) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:899]: Having `assertUpdate(admin, "CREATE VIEW test_view_authorization.bar AS SELECT current_user AS user");` would not decrease the coverage, but surely would simplify the test.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 6]
```diff
@@ -853,6 +853,126 @@ public void testTableAuthorizationForRole()
...
+                .setCatalog(getSession().getCatalog().get())
+                .setIdentity(Identity.forUser("hive").withRole("hive", new SelectedRole(ROLE, Optional.of("admin"))).build())
+                .build();
+
+        Session alice = testSessionBuilder()
+                .setCatalog(getSession().getCatalog().get())
+                .setIdentity(Identity.forUser("alice").build())
+                .build();
+
+        assertUpdate(admin, "CREATE SCHEMA test_view_authorization");
```
Comment (kokosing) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:896]: use `io.prestosql.testing.sql.TestTable#randomTableSuffix` so names are random

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 7]
```diff
@@ -853,6 +853,126 @@ public void testTableAuthorizationForRole()
...
+        assertQueryFails(alice, "SELECT * FROM test_view_authorization.bar", "Access Denied: Cannot select from table test_view_authorization.foo");
+
+        assertUpdate(alice, "ALTER VIEW test_view_authorization.bar SET AUTHORIZATION admin");
+        assertUpdate(admin, "DROP VIEW test_view_authorization.bar");
+        assertUpdate(admin, "DROP TABLE test_view_authorization.foo");
+        assertUpdate(admin, "DROP SCHEMA test_view_authorization");
+    }
+
+    @Test
+    public void testViewAuthorizationSecurityInvoker()
```
Comment (kokosing) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:913]: same comments as above

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestSqlStandardAccessControlChecks.java [hunk 8]
```diff
@@ -221,6 +221,15 @@ public void testAccessControlSetTableAuthorization()
...
     }
 
+    @Test(groups = {AUTHORIZATION, PROFILE_SPECIFIC_TESTS})
+    public void testAccessControlSetViewAuthorization()
+    {
+        assertThat(() -> bobExecutor.executeQuery(format("ALTER VIEW %s SET AUTHORIZATION bob", viewName)))
+                .failsWithMessage(format("Access Denied: Cannot set authorization for view default.%s to USER bob", viewName));
+        aliceExecutor.executeQuery(format("ALTER VIEW %s SET AUTHORIZATION bob", viewName));
+        bobExecutor.executeQuery(format("ALTER VIEW %s SET AUTHORIZATION alice", viewName));
+    }
```
Comment (kokosing) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestSqlStandardAccessControlChecks.java:231]: Please add a test with hive view (where views is defined in Hive).

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestSqlStandardAccessControlChecks.java [hunk 9]
```diff
@@ -221,6 +221,15 @@ public void testAccessControlSetTableAuthorization()
         bobExecutor.executeQuery(format("ALTER TABLE %s SET AUTHORIZATION alice", tableName));
     }
 
+    @Test(groups = {AUTHORIZATION, PROFILE_SPECIFIC_TESTS})
+    public void testAccessControlSetViewAuthorization()
+    {
+        assertThat(() -> bobExecutor.executeQuery(format("ALTER VIEW %s SET AUTHORIZATION bob", viewName)))
+                .failsWithMessage(format("Access Denied: Cannot set authorization for view default.%s to USER bob", viewName));
+        aliceExecutor.executeQuery(format("ALTER VIEW %s SET AUTHORIZATION bob", viewName));
```
Comment (kokosing) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestSqlStandardAccessControlChecks.java:229]: There is missing assertion that would prove that authorization was actually set. Below line could pass from other reasons too.

File: presto-plugin-toolkit/src/test/java/io/prestosql/plugin/base/security/TestFileBasedSystemAccessControl.java [hunk 10]
```diff
@@ -588,13 +597,29 @@ public void testCheckCanSetTableAuthorizationForOwner()
         accessControl.checkCanSetViewAuthorization(ALICE, new CatalogSchemaTableName("some-catalog", "aliceschema", "test"), new PrestoPrincipal(PrincipalType.USER, "some_user"));
     }
 
+    @Test
```
Comment (kokosing) [inline presto-plugin-toolkit/src/test/java/io/prestosql/plugin/base/security/TestFileBasedSystemAccessControl.java:600]: I guess this belongs to different commit.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java (hunk 5 above)
Comment (skrzypo987) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:899]: Too bad it does not work. Executing current_user seems to always return the user being logged in, not the one executing the query,

Comment (kokosing) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:899]: That is the way it should work for DEFINER.

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestSqlStandardAccessControlChecks.java (hunk 8 above)
Comment (skrzypo987) [reply presto-product-tests/src/main/java/io/prestosql/tests/hive/TestSqlStandardAccessControlChecks.java:234]: I get a `Hive views are not supported` when using hive views. I don't think they are supported in this setup.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java (hunk 5 above)
Comment (skrzypo987) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: It does not.

**Comment by skrzypo987** [comment]
@kokosing Added a product test with hive views as a separate commit

File: presto-parser/src/main/java/io/prestosql/sql/tree/SetAuthorizationStatement.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
+/*
```
Comment (kokosing) [inline presto-parser/src/main/java/io/prestosql/sql/tree/SetAuthorizationStatement.java:1]: Update commit message to:
```
Extract SetAuthorizationStatement class
```

**Review by kokosing — APPROVED** [review]
Please ping when automation passes.

**Comment by kokosing** [comment]
CI hit: https://github.com/prestosql/presto/issues/5422

**Comment by kokosing** [comment]
Merged, thanks!
